src/Compilers/VisualBasic/Portable/Symbols/Tuples/TupleTypeSymbol.vb

```diff
@@ -0,0 +1,987 @@
+            Dim num As Integer = TupleTypeSymbol.NumberOfValueTuples(cardinality, arity)
+            Dim wellKnownType As NamedTypeSymbol = compilation.GetWellKnownType(TupleTypeSymbol.GetTupleType(arity))
+            Binder.ReportUseSiteError(diagnostics, syntax, wellKnownType)
+
+            If num > 1 Then
+                Dim wellKnownType2 As NamedTypeSymbol = compilation.GetWellKnownType(TupleTypeSymbol.GetTupleType(TupleTypeSymbol.RestPosition))
+                Binder.ReportUseSiteError(diagnostics, syntax, wellKnownType2)
+            End If
+        End Sub
+
+        Private Shared Function GetTupleType(arity As Integer) As WellKnownType
+            If arity > TupleTypeSymbol.RestPosition Then
+                Throw ExceptionUtilities.Unreachable
+            End If
+
+            Return TupleTypeSymbol.tupleTypes(arity - 1)
+        End Function
+
+        Friend Shared Function GetTupleCtor(arity As Integer) As WellKnownMember
+            If arity > TupleTypeSymbol.RestPosition Then
+                Throw ExceptionUtilities.Unreachable
+            End If
+
+            Return TupleTypeSymbol.tupleCtors(arity - 1)
+        End Function
+
+        Friend Shared Function GetTupleTypeMember(arity As Integer, position As Integer) As WellKnownMember
+            Return TupleTypeSymbol.tupleMembers(arity - 1)(position - 1)
+        End Function
+
+        Friend Shared Function TupleMemberName(position As Integer) As String
+            Return "Item" & position
+        End Function
+
+        ' PROTOTYPE: case sensitivity
+        Private Shared Function IsElementNameForbidden(name As String) As Boolean
+            Return name = "CompareTo" OrElse name = "Deconstruct" OrElse name = "Equals" OrElse name = "GetHashCode" OrElse name = "Rest" OrElse name = "ToString"
+        End Function
+
+        ' PROTOTYPE: case sensitivity
+        Friend Shared Function IsElementNameReserved(name As String) As Integer
+            Dim result As Integer
+            If TupleTypeSymbol.IsElementNameForbidden(name) Then
+                result = 0
+            Else
+                If name.StartsWith("Item", StringComparison.Ordinal) Then
+                    Dim s As String = name.Substring(4)
+                    Dim num As Integer
+
+                    If Integer.TryParse(s, num) Then
+                        If num > 0 AndAlso String.Equals(name, TupleTypeSymbol.TupleMemberName(num), StringComparison.Ordinal) Then
+                            result = num
+                            Return result
+                        End If
+                    End If
+                End If
+                result = -1
+            End If
+            Return result
+        End Function
+
+        Friend Shared Function GetWellKnownMemberInType(type As NamedTypeSymbol, relativeMember As WellKnownMember) As Symbol
+            Debug.Assert(relativeMember >= WellKnownMember.System_Math__RoundDouble AndAlso relativeMember < WellKnownMember.Count)
```

Comment:
done in both VB and C#

---
In reply to: [71562273](https://github.com/dotnet/roslyn/pull/12596#discussion_r71562273) [](ancestors = 71562273)